7_drivers_munge/src/GCM_driver_nc_utils.R

```diff
@@ -3,25 +3,99 @@
+  # Get vectors of cell centroid latitudes and longitudes
+  cell_centroid_lats <- cell_coords %>% pull(lat)
+  cell_centroid_lons <- cell_coords %>% pull(lon)
+
+  # Set up attributes for NetCDF that are independent of variables
+  data_time_units <- "days since 1970-01-01 00:00:00"
+  data_attributes <- list('title' = global_att)
+  data_attributes <- append(data_attributes, grid_params) # add grid parameters to list of global attributes
+  data_coordvar_long_names <- list(instance = "identifier for GCM grid cell", time = "date",
+                              lat = "WGS84 latitude of GCM grid cell centroid", lon = "WGS84 longitude of GCM grid cell centroid",
+                              alt = "NULL")
+
+  # Pivot data to long format to set up for filtering by variable
+  gcm_data_long <- gcm_data %>%
+    pivot_longer(cols = -c(time, cell), names_to = "variable", values_to = "value") %>%
+    arrange(cell)
+
+  # Loop over data variables to populate NetCDF using ncdfgeom::write_timeseries_dsg()
+  for (variable in vars_info$var_name) {
+    # Filter long-format data to selected variable
+    variable_metadata <- vars_info %>% filter(var_name == variable)
+    var_data <- gcm_data_long %>%
+      filter(variable == !!variable) %>%
+      pivot_wider(id_cols = c("time"), names_from = cell, values_from = value) %>%
+      select(-time) %>%
+      as.data.frame() %>%
+      setNames(gcm_cells)
+
+    # Pull attributes for selected variable
+    var_data_unit <- rep(variable_metadata$units, length(gcm_cells))
+    var_data_prec <- variable_metadata$precision
+    var_data_metadata <- list(name = variable_metadata$var_name, long_name = variable_metadata$longname)
+
+    # Check if nc file already exists, and therefore should be added to,
+    # or if it needs to be created
+    var_add_to_existing <- ifelse(file.exists(nc_file), TRUE, FALSE)
+
+    # write this variable to the netcdf file:
+    # ncdfgeom::write_timeseries_dsg() function documentation:
+    # https://github.com/USGS-R/ncdfgeom/blob/master/R/write_timeseries_dsg.R
+    ncdfgeom::write_timeseries_dsg(nc_file,
+                         instance_names = gcm_cells,
+                         lats = cell_centroid_lats,
+                         lons = cell_centroid_lons,
+                         alts = NA,
+                         times = gcm_dates,
+                         data = var_data,
+                         data_unit = var_data_unit,
+                         data_prec = var_data_prec,
+                         data_metadata = var_data_metadata,
+                         time_units = data_time_units,
+                         instance_dim_name = gcm_cells_dim_name,
+                         dsg_timeseries_id = gcm_cells_dim_name,
+                         attributes = data_attributes,
+                         coordvar_long_names = data_coordvar_long_names,
+                         add_to_existing = var_add_to_existing,
+                         overwrite = overwrite)
+  }
```

Comment:
One thing to consider is whether you'd benefit from compression of this file, or if it is necessary ([see system calls here](https://github.com/jread-usgs/lakesurf-data-release/blob/main/src/ncdf_utils.R#L85))

Reply:
Yes, good point. I'd seen that in your original code, and figured we'd revisit once this code was up and running. Lindsay and I checked and right now the raw feather files total 1.91 gb and the netCDF fiels total 1.24 gb, so it's a reduction but not a huge reduction. How much more of a reduction did compression get you in the `lakesurf-data-release`?

Reply:
Hmmm...I checked on this one with this
```
ncks -h --fl_fmt=netcdf4 --cnk_plc=g3d --cnk_dmn time,10 --ppc Rain=.3#Snow=.3#AirTemp=.2#RelHum=.1#Shortwave=.1#Longwave=.1#WindSpeed=.3 7_GCM_MRI.nc compressed_MRI_out.nc
```
And `7_GCM_MRI.nc` (original file) is 223.4 MB and `compressed_MRI_out.nc` is 72 MB for me. The ".3" and ".1" are the different precisions I set for the different variables as a starting point. Otherwise I was using the existing code from that other repo.

Reply:
@jread-usgs - I've added in the code to write to temporary files and then compress those, but can't test it on my machine because I don't have the command-line programs to do so. Is [NCO](https://github.com/nco/nco) what I need?

Reply:
Yes, nco and the netcdf libraries should do it, but I bet you already have the netcdf libraries installed. I use homebrew to install these on mac, but not sure what to recommend for windows.

Reply:
Hmm I installed via conda-forge, which seems to be what is recommended for Windows [here](http://nco.sourceforge.net/). I can now compress the files with the ncks command in an Anaconda Prompt shell, but not from R, within the pipeline. 🙁.

For me the files go down from 213mb to 68.7mb

Reply:
@dblodgett-usgs got install recommendations for Hayley in this case?

Reply:
Per [the README.md of this R package wrapper of NCO](https://github.com/markpayneatwork/ClimateOperators), which I've not yet attempted to install, just stumbled upon, I've been trying to make the conda installation of NCO [available in my system path](https://github.com/markpayneatwork/ClimateOperators/blob/master/README.md?plain=1#L15), so that I can call it from R using system commands. I'm a bit out of my depth here, but when I added `C:\Users\hcorson-dosch\Anaconda3\pkgs\nco-4.8.1-habc9422_100\Library\bin` to my PATH environment variable, I now no longer get a `ncks: command not found` error in Git Bash, but I do get this error: 
```
hcorson-dosch@IGSARMEWLTHCOR MINGW64 ~/Documents/R/lake-temperature-model-prep/7_drivers_munge/out (gcm_driver_data_munge_pipeline)
$ ncks
C:/Users/hcorson-dosch/Anaconda3/pkgs/nco-4.8.1-habc9422_100/Library/bin/ncks.exe: error while loading shared libraries: api-ms-win-crt-environment-l1-1-0.dll: cannot open shared object file: No such file or directory
```

Reply:
Okay hit a whole series of missing .dll files, most of which I was able to copy to the directory of the `ncks.exe` from `C:\Users\hcorson-dosch\Anaconda3\Library\bin`, but not having luck finding `udunits2.dll` which it now can't find.

If I now attempt to run the pipeline and compressed the files through RStudio, I get these error messages:
![image](https://user-images.githubusercontent.com/54007288/152015110-d8cb85fa-e5d2-44df-b115-a2836e11c377.png)
![image](https://user-images.githubusercontent.com/54007288/152015153-ecdec656-e3c2-4137-99a4-c1882fb38193.png)

Reply:
Alright I've installed `udunits2` from CRAN, and copied `udunits2.dll` to `C:\Users\hcorson-dosch\Anaconda3\pkgs\nco-4.8.1-habc9422_100\Library\bin` from `C:\Users\hcorson-dosch\Anaconda3\pkgs\udunits2-2.2.27.27-h892ecd3_3\Scripts`, which has fixed that first error.

Reply:
Not a comment to help out with this error (because I have no idea), but let's make sure we note these external libraries somewhere and maybe log an issue to add a helpful error message if you are missing them in the future.

Reply:
Another thought: while compression is important in the end and a general good idea/best practice, let's also keep in mind practicality for our timeline. We want to get some results up by the end of the week and maybe we can't get a compressed NetCDF by then. In order to keep moving forward, maybe we log an issue for adding file compression, link to this thread, and then keep moving forward with an MVP?

Reply:
Alright just added another commit that makes `compression` a T/F argument to the `generate_gcm_nc()` function. If it is set to TRUE and fails, the function will throw an error. Note: I tested but didn't use a tryCatch because if NCO isn't set up R throws a system error, but not a console error, so it isn't caught by TryCatch. So instead I'm using a simple check to see if the output compressed file was created, and, if not, throwing an error.

Reply:
@jread-usgs - can you confirm that the compression code works on your end? Note you'll need to change the `compression` argument in the target recipe for `gcm_nc`, to `TRUE`.

If it does work, I'll add another commit deleting my workaround for the `glm_ready_gcm_data_feather` files, and then @lindsayplatt, could you test the code on your end (with `compression` = `FALSE`) to make sure the pipeline builds using the actual pipeline built `glm_ready_gcm_data_feather` files, rather than those I manually provided?

Reply:
For NetCDF and nco, I think you can install using chocolaty. I would say run with Conda if it works, but have never had good luck using Conda-installed-dependencies outside of a Conda environment. 

I thought I'd installed it with chocolatey -- but now I can't find that source. 

I think you could probably just install the precompiled binary NetCDF (netCDF4+DAP) and use nccopy. nccopy does chunking and deflation. 

`nco` is not available via precompiled binary AFAIK, and it required for that --ppc lossy compression scheme. I use WSL linux or a dockerized (linux) NetCDF to do that kind of stuff rather than messing with getting it compiled and on my path in Windows.

Reply:
I didn't realize this was such a headache on Windows...

Reply:
Theoretically, the conda-forge solution should work... I have just never taken the time to figure out how to get it working correctly.